rubicon/objc/types.py

```diff
@@ -18,17 +18,194 @@
 __arm__ = (_any_arm and not __LP64__)
 
-PyObjectEncoding = b'{PyObject=@}'
 
-
-def encoding_for_ctype(vartype):
-    typecodes = {
-        c_char: b'c', c_int: b'i', c_short: b's', c_long: b'l', c_longlong: b'q',
-        c_ubyte: b'C', c_uint: b'I', c_ushort: b'S', c_ulong: b'L', c_ulonglong: b'Q',
-        c_float: b'f', c_double: b'd', c_bool: b'B', c_char_p: b'*', c_void_p: b'@',
-        py_object: PyObjectEncoding
-    }
-    return typecodes.get(vartype, b'?')
+_type_to_ctype_map = {
+    int: c_int,
+    float: c_float,
+    bool: c_bool,
+    # This is not really correct - Python strings are Unicode - but we need to keep this for compatibility.
+    str: c_char_p,
+    bytes: c_char_p,
+}
+
+def type_to_ctype(tp):
+    """Convert the given type to a ctypes type.
+    This conversion is applied to types used in objc_method signatures, objc_ivar types, etc.
+    This translates Python built-in types and rubicon.objc classes to their ctypes equivalents.
+    Unregistered types (including things that are already ctypes types) are returned unchanged.
+    """
+    
+    return _type_to_ctype_map.get(tp, tp)
+
+def register_type_to_ctype(tp, ctype):
+    """Register a conversion from a type to a ctypes type."""
+    
+    _type_to_ctype_map[tp] = ctype
+
+def unregister_type_to_ctype(tp):
+    """Unregister a conversion from a type to a ctypes type."""
+    
+    del _type_to_ctype_map[tp]
+
+def get_type_to_ctype_map():
+    """Get a copy of all currently registered type-to-ctype conversions as a mapping."""
+    
+    return dict(_type_to_ctype_map)
+
+_encoding_to_ctype_map = {}
+_ctype_to_encoding_map = {}
+
+def encoding_to_ctype(encoding):
+    """Return ctypes type for an encoded Objective-C type."""
+    
+    # Remove qualifiers, as documented in Table 6-2 here:
+    # https://developer.apple.com/library/prerelease/content/documentation/Cocoa/Conceptual/ObjCRuntimeGuide/Articles/ocrtTypeEncodings.html
+    encoding = encoding.lstrip(b"NORVnor")
+    
+    try:
+        # Look up simple type encodings directly
+        return _encoding_to_ctype_map[encoding]
+    except KeyError:
+        if encoding[0:1] == b'^':
+            try:
+                # Try to resolve pointer types recursively
+                target = encoding_to_ctype(encoding[1:])
+            except ValueError:
+                # For unknown pointer types, fall back to c_void_p (this is not ideal, but at least it works instead of erroring)
+                return c_void_p
+            else:
+                return POINTER(target)
+        else:
+            raise ValueError('Unknown encoding: %s' % (encoding,))
+
+def ctype_to_encoding(ctype):
+    """Return the Objective-C type encoding for the given ctypes type."""
+    
+    try:
+        return _ctype_to_encoding_map[ctype]
+    except KeyError:
+        raise ValueError('No type encoding known for ctype {}'.format(ctype))
+
+def register_preferred_encoding(encoding, ctype):
```

Comment:
While we're here bikeshedding names... I was thinking about renaming this to `register_encoding`, and the current `register_encoding` to `register_alternative_encoding`, or something similar.

The reason being that I later want to make struct types be created automatically for unknown struct encodings. In that case, when a user registers a custom struct type, they probably want it to take priority over the auto-generated one, so it makes more sense for overwriting to be the default behavior.

On the other hand, there is also the possibility of users registering custom types that correspond to integers, such as enums. In that case, the correct behavior would be to *not* overwrite existing encodings, otherwise every integer would be converted to that enum... I think this case is much less likely though, because we already "unwrap" enum constants automatically when passed to an Objective-C method. There's also `enum.IntEnum` and `enum.IntFlag` which can be used for int-compatible enums.